### ISSUE TITLE
Bug with yum and with_items when installing RPMs

### ISSUE THREAD

**Issue by beenje** [ansible/ansible-modules-core#843]
##### Issue Type:

Bug Report
##### Ansible Version:

```
$ ansible --version
ansible 1.9 (devel af59e108a5) last updated 2015/02/24 22:01:43 (GMT +200)
  lib/ansible/modules/core: (detached HEAD 706be79f8f) last updated 2015/02/24 22:01:47 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD 2ffd7235e5) last updated 2015/02/24 22:01:51 (GMT +200)
  v2/ansible/modules/core: (detached HEAD 34784b7a61) last updated 2015/02/24 22:01:55 (GMT +200)
  v2/ansible/modules/extras: (detached HEAD 312d59061c) last updated 2015/02/24 22:01:59 (GMT +200)
  configured module search path = None
```
##### Environment:

Ansible running on OS X, managing CentOS 6.5
##### Summary:

When installing a list of RPMs using with_items, the full path is only kept for the first item and hence installation fails for other items in the list.
##### Steps To Reproduce:

```

---
- hosts: all
  vars:
    mongodb_packages:
      - mongodb-org-tools-2.6.0-1.x86_64.rpm
      - mongodb-org-shell-2.6.0-1.x86_64.rpm
      - mongodb-org-mongos-2.6.0-1.x86_64.rpm
      - mongodb-org-server-2.6.0-1.x86_64.rpm
      - mongodb-org-2.6.0-1.x86_64.rpm
  sudo: yes
  tasks:
    - name: copy mongodb packages
      copy: src={{item}} dest=/tmp/
      with_items: mongodb_packages
    - name: install mongodb packages
      yum: name="/tmp/{{item}}" disablerepo=* state=present
      with_items: mongodb_packages
```
##### Expected Results:

The expected results is that all RPMs should be installed. The following list should be passed to yum:

```
/tmp/mongodb-org-tools-2.6.0-1.x86_64.rpm, /tmp/mongodb-org-shell-2.6.0-1.x86_64.rpm, /tmp/mongodb-org-mongos-2.6.0-1.x86_64.rpm, /tmp/mongodb-org-server-2.6.0-1.x86_64.rpm, /tmp/mongodb-org-2.6.0-1.x86_64.rpm
```
##### Actual Results:

Only the first RPM is installed because the full path is not passed with other items and the files can't be found :

```
name=/tmp/mongodb-org-tools-2.6.0-1.x86_64.rpm,mongodb-org-shell-2.6.0-1.x86_64.rpm,mongodb-org-mongos-2.6.0-1.x86_64.rpm,mongodb-org-server-2.6.0-1.x86_64.rpm,mongodb-org-2.6.0-1.x86_64.rpm
```

See full log:

```
TASK: [install mongodb packages] **********************************************
<127.0.0.1>
<127.0.0.1> state=present disablerepo=* name=/tmp/mongodb-org-tools-2.6.0-1.x86_64.rpm,mongodb-org-shell-2.6.0-1.x86_64.rpm,mongodb-org-mongos-2.6.0-1.x86_64.rpm,mongodb-org-server-2.6.0-1.x86_64.rpm,mongodb-org-2.6.0-1.x86_64.rpm
<127.0.0.1> IdentityFile=/Users/benjamin/Dev/vagrant/.vagrant/machines/default/virtualbox/private_key ConnectTimeout=10 PasswordAuthentication=no KbdInteractiveAuthentication=no User=vagrant ControlPath=/Users/benjamin/.ansible/cp/ansible-ssh-%h-%p-%r PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey ControlMaster=auto Port=2222 ControlPersist=60s
<127.0.0.1>
<127.0.0.1> IdentityFile=/Users/benjamin/Dev/vagrant/.vagrant/machines/default/virtualbox/private_key ConnectTimeout=10 'sudo -k && sudo -H -S -p "[sudo via ansible, key=ydkqokftkwaqihugjdvyfkydpfctdnul] password: " -u root /bin/sh -c '"'"'echo SUDO-SUCCESS-ydkqokftkwaqihugjdvyfkydpfctdnul; LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python -tt /home/vagrant/.ansible/tmp/ansible-tmp-1424814194.07-262712224273338/yum; rm -rf /home/vagrant/.ansible/tmp/ansible-tmp-1424814194.07-262712224273338/ >/dev/null 2>&1'"'"'' PasswordAuthentication=no KbdInteractiveAuthentication=no User=vagrant ControlPath=/Users/benjamin/.ansible/cp/ansible-ssh-%h-%p-%r PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey ControlMaster=auto Port=2222 ControlPersist=60s
failed: [machine] => (item=mongodb-org-tools-2.6.0-1.x86_64.rpm,mongodb-org-shell-2.6.0-1.x86_64.rpm,mongodb-org-mongos-2.6.0-1.x86_64.rpm,mongodb-org-server-2.6.0-1.x86_64.rpm,mongodb-org-2.6.0-1.x86_64.rpm) => {"changed": false, "failed": true, "item": "mongodb-org-tools-2.6.0-1.x86_64.rpm,mongodb-org-shell-2.6.0-1.x86_64.rpm,mongodb-org-mongos-2.6.0-1.x86_64.rpm,mongodb-org-server-2.6.0-1.x86_64.rpm,mongodb-org-2.6.0-1.x86_64.rpm", "rc": 0, "results": []}
msg: No Package file matching 'mongodb-org-shell-2.6.0-1.x86_64.rpm' found on system

FATAL: all hosts have already failed -- aborting
```

**Comment by bcoca** [comment]
this is how with_items currently works with yum and other specific modules (apt, zypper, etc), there  is an optimization for it to only run one time for the full list passed in with_ , which makes it add to name="{{item}}',  anything else in name= field will only work for the first item in the list.

the current work around would be to add the path to each item in the list

**Comment by beenje** [comment]
Thanks for the quick feedback.
Do you have any pointer if I'm interested in trying to fix that? What's the best way to debug an ansible module?

**Comment by bcoca** [comment]
The fix is not trivial as this is not module code but has to do with the runner object, which  the one that takes, parses the arguments and then feeds them to the module for execution. This affects other modules and all would have to be tested.

I do have it on my list to make it smarter and only run with 
name={{item}} or be able to use pre/postfixes into every part of the list.

**Comment by ppmathis** [comment]
Just wanted to add something, as I struggled with the same bug/issue: You can easily create your own filter plugin as a workaround. Whenever Ansible is missing something, I just add my own module into the mix and it works like a charm. For example, the following filter module adds **items_prepend()** and **items_append()**:

``` python
from ansible import errors

def items_prepend(array, prepend):
    if isinstance(array, list):
        return [prepend + item for item in array]
    else:
        raise errors.AnsibleFilterError('items_prepend() can only be used on arrays: %s' % str(array))

def items_append(array, append):
    if isinstance(array, list):
        return [item + append for item in array]
    else:
        raise errors.AnsibleFilterError('items_append() can only be used on arrays: %s' % str(array))

class FilterModule(object):
    def filters(self):
        return {
            'items_prepend': items_prepend,
            'items_append': items_append
        }
```

You can use it like this:

``` yaml
with_items: '{{ common.software.repositories|items_prepend("/usr/local/src/repositories/") }}'
```

As the ansible configuration file (_ansible.cfg_) allows you to specify the path to your plugins, you can also easily integrate all those custom modules into your repository.

**Comment by kapilt** [comment]
just ran into this as well, for folks having to install packages from a set of files, this is a serious annoyance and inconsistency. there's tons of other modules that with_items and prefixes work with fine ie (copy).

**Comment by mkoshelev** [comment]
Just my 2c - you can use with_nested instead of with_items as a workaround. Put your original list as the first one for with_nested and one-item dummy list as second - and replace {{ item }} with {{ item[0] }} :

```
- name: Install some packages
  yum: name="http://URL/pkgname-{{ item[0] }}-suffix.rpm" state=present
  with_nested:
   - [ 'one', 'two' ]
   - [ True ]
```

**Comment by Rucknar** [comment]
Also hitting the same issue, i can't think of a use-case where you would want it to behave like this. As @kapilt says, if this behaves differently with different modules then this is a bug.

**Comment by bcoca** [comment]
so this is fixed in 2.0 by making the optimization much smarter and keeping any prefix/postfix existing in name.
